### DISCUSSION TITLE
How to deal with nan value in the buffer?

### DISCUSSION THREAD

**Discussion by liudiyi-mechmind** [halide/Halide#8138] [Q&A] [answered]
Hi, I'm a beginner in Halide. This work is truly remarkable.

My question is how to deal with nan value in the buffer.
Let's say I am going to write a blur1x3, but there may be nan values in the buffer.
here is my code.
`average(x, y, c) = select(is_nan(input(x, y, c)), NAN, average3(input(x - 1, y, c), input(x, y, c), input(x + 1, y, c)));
 
Expr average3(Expr a, Expr b, Expr c)
  {
      return select(!is_nan(a) && !is_nan(c), (a + b + c) / (float)3.0,
                    select(is_nan(a) && is_nan(c), b,
                           select(is_nan(a) && !is_nan(c), (c + b) / (float)2.0,
                                  (a + b) / (float)2.0)));
  }
`
But when it comes 1 * 5 or 1 * 7, the codes are getting complicated in this way.
Is there any good idea to help me write code for 1*5?
Thank you in advance.

**Comment by zvookin** [comment, marked as answer]
Generally, unless one turns on `strict_float`, NaN is not allowed. If `strict_float` is in effect, NaNs should propagate through most mathematical operations. Comparisons involving NaN always fail per their specification. There are of course multiple kinds of NaN and theoretically `strict_float` should obey IEEE 754 rules, but we don't have test coverage for this and I wouldn't be at all surprised if the exact rules are not always followed. That said one should get a NaN result for calculations using NaN inputs.

`strict_float` can either be specified as a target flag or used as an operator around specific expressions.

As a practical matter, for performant pipelines in many domains, a NaN in the input is an error and it is best to to preprocess the input to filter them if they can occur.

**Reply by liudiyi-mechmind** [reply]
Thank you very much for your prompt response.

Returning the NaN value as the result or performing some preprocessing seems to be able to address this issue. In fact, I am working in the field of high-speed 3D measurement using scanned data. The depth maps we capture often contain NaN values due to factors such as shadows. I will explore possible workarounds for this issue.

Thank you once again for your helpful response.

**Comment by abadams** [comment]
I recommend having a separate uint8 buffer that's 1 where there is valid data and 0 elsewhere. You then do the convolution by doing a weighted convolution using the separate buffer for the weights. You'll need to divide by the local sum of the weights to normalize.
